public/apps/configuration/panels/role-edit.tsx

```diff
@@ -47,24 +58,212 @@ const TITLE_TEXT_DICT = {
   create: 'Create Role',
   edit: 'Edit Role',
   duplicate: 'Duplicate Role',
-}
+};
 
 type OptionSeletion = EuiComboBoxOptionOption[];
+type FieldLevelSecurityMethod = 'exclude' | 'include';
+
+interface RoleIndexPermissionStateClass {
+  indexPatterns: OptionSeletion;
+  docLevelSecurity: string;
+  fieldLevelSecurityMethod: FieldLevelSecurityMethod;
+  fieldLevelSecurityFields: OptionSeletion;
+  maskedFields: OptionSeletion;
+  allowedActions: OptionSeletion;
+}
+
+function getEmptyIndexPermission(): RoleIndexPermissionStateClass {
+  return {
+    indexPatterns: [],
+    docLevelSecurity: '',
+    fieldLevelSecurityMethod: 'exclude',
+    fieldLevelSecurityFields: [],
+    maskedFields: [],
+    allowedActions: [],
+  };
+}
+
+/**
+ * Identify the method is whether exclude or include.
+ * @param fieldLevelSecurityRawFields fields fetched from backend
+ * ["~field1", "~field2"] => exclude
+ * ["field1", "field2"] => include
+ */
+function getFieldLevelSecurityMethod(fieldLevelSecurityRawFields: string[]): FieldLevelSecurityMethod {
+  // Leading ~ indicates exclude.
+  return fieldLevelSecurityRawFields.some((s: string) => s.startsWith('~')) ? 'exclude' : 'include';
+}
+
+/**
+ * Remove the leading ~ which indicates exclude and convert to combo box option.
+ * @param fieldLevelSecurityRawFields fields fetched from backend
+ * ["~field1", "~field2"] => ["field1", "field2"]
+ * ["field1", "field2"] => ["field1", "field2"]
+ */
+function getFieldLevelSecurityFields(fieldLevelSecurityRawFields: string[]): OptionSeletion {
+  return fieldLevelSecurityRawFields.map((s: string) => s.replace(/^~/, '')).map(stringToComboBoxOption)
+}
+
+function buildIndexPermissionState(
+  indexPerm: RoleIndexPermission[]
+): RoleIndexPermissionStateClass[] {
+  return indexPerm.map(perm => ({
+    indexPatterns: perm.index_patterns.map(stringToComboBoxOption),
+    docLevelSecurity: perm.dls,
+    fieldLevelSecurityMethod: getFieldLevelSecurityMethod(perm.fls),
+    fieldLevelSecurityFields: getFieldLevelSecurityFields(perm.fls),
+    maskedFields: [],
+    allowedActions: [],
+  }));
+}
+
+const FIELD_LEVEL_SECURITY_PLACEHOLDER = `{
+    "bool": {
+        "must": {
+            "match": {
+                "genres": "Comedy"
+            }
+        }
+    }
+}`;
+
+function generateIndexPermissionPanels(
+  indexPermissions: RoleIndexPermissionStateClass[],
+  permisionOptionsSet: EuiComboBoxOptionOption[],
+  setRoleIndexPermission: Dispatch<SetStateAction<RoleIndexPermissionStateClass[]>>
+) {
+  const panels = indexPermissions.map((perm, index) => {
+    const onValueChangeHandler = (field: string) =>
+      updateElementInArrayHandler(setRoleIndexPermission, [index, field]);
 
-function buildPermissionOptions(optionsList: string[]) {
-  return optionsList.map(e => ({ label: e }));
+    const onCreateOptionHandler = (field: string) =>
+      appendOptionToComboBoxHandler(setRoleIndexPermission, [index, field]);
+
+    return (
```

Comment:
Can we break down the return to some sub render functions for better readability?

Reply:
+1

Reply:
Good idea, broke.